src/intptrcast.rs

```diff
@@ -0,0 +1,103 @@
+use std::cell::{Cell, RefCell};
+
+use rustc::mir::interpret::{AllocId, Pointer, InterpResult};
+use rustc_mir::interpret::Memory;
+use rustc_target::abi::Size;
+
+use crate::stacked_borrows::Tag;
+use crate::Evaluator;
+
+pub type MemoryExtra = RefCell<GlobalState>;
+
+#[derive(Clone, Debug, Default)]
+pub struct AllocExtra {
+    base_addr: Cell<Option<u64>>
+}
+
+#[derive(Clone, Debug)]
+pub struct GlobalState {
+    /// This is used as a map between the address of each allocation and its `AllocId`.
+    /// It is always sorted
+    pub int_to_ptr_map: Vec<(u64, AllocId)>,
+    /// This is used as a memory address when a new pointer is casted to an integer. It
+    /// is always larger than any address that was previously made part of a block.
+    pub next_base_addr: u64,
+}
+
+impl Default for GlobalState {
+    // FIXME: Query the page size in the future
+    fn default() -> Self {
+        GlobalState {
+            int_to_ptr_map: Vec::default(),
+            next_base_addr: 2u64.pow(16)
+        }
+    }
+}
+
+impl<'mir, 'tcx> GlobalState {
+    pub fn int_to_ptr(
+        base_addr: u64,
```

Comment:
No this shouldn't be `base_addr`, it's not a base address after all. "base address" refers to the address where some allocation starts.